src/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs

```diff
@@ -0,0 +1,146 @@
+        {
+            if (protocol == null)
+            {
+                throw new ArgumentNullException(nameof(protocol));
+            }
+
+            if (protocol.Length == 0)
+            {
+                throw new ArgumentException(SR.net_ssl_app_protocol_empty, nameof(protocol));
+            }
+
+            if (copy)
+            {
+                _protocol = new byte[protocol.Length];
+                Array.Copy(protocol, _protocol, protocol.Length);
+            }
+            else
+            {
+                _protocol = protocol;
+            }
+
+            _readOnlyProtocol = new ReadOnlyMemory<byte>(_protocol);
+        }
+
+        public SslApplicationProtocol(byte[] protocol) : this(protocol, true) { }
+
+        public SslApplicationProtocol(string protocol) : this(s_utf8.GetBytes(protocol), false) { }
+
+        public ReadOnlyMemory<byte> Protocol
+        {
+            get => _readOnlyProtocol;
+        }
+
+        public bool Equals(SslApplicationProtocol other)
+        {
+            if (_protocol == null && other._protocol == null)
+                return true;
+
+            if (_protocol == null || other._protocol == null)
+                return false;
+
+            if (_protocol.Length != other._protocol.Length)
+                return false;
+
+            for (int i = 0; i < _protocol.Length; i++)
+            {
+                if (_protocol[i] != other._protocol[i])
+                    return false;
+            }
+
+            return true;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is SslApplicationProtocol protocol)
+            {
+                return Equals(protocol);
+            }
+
+            return false;
+        }
+
+        public override int GetHashCode()
+        {
+            int hash = 0;
+            for (int i = 0; i < _protocol.Length; i++)
+            {
+                hash = ((hash << 5) + hash) ^ _protocol[i];
+            }
+
+            return hash;
+        }
+
+        public override string ToString()
+        {
+            try
+            {
+                if (_protocol == null)
+                {
+                    return null;
+                }
+
+                return s_utf8.GetString(_protocol);
+            }
+            catch
+            {
+                // In case of decoding errors, return the byte values as hex string.
+                int byteCharsLength = _protocol.Length * 5;
+                char[] byteChars = new char[byteCharsLength];
+                int index = 0;
+
+                for (int i = 0; i < byteCharsLength; i += 5)
+                {
+                    byte b = _protocol[index++];
+                    byteChars[i] = '0';
+                    byteChars[i + 1] = 'x';
+                    byteChars[i + 2] = GetHexValue(b / 16);
+                    byteChars[i + 3] = GetHexValue(b % 16);
+                    byteChars[i + 4] = ' ';
+                }
```

Comment:
Why invent a new form here?  `BitConverter.ToString(ReadOnlySpan<byte>)` was presumably added during the byte[] => (ReadOnly)Span-ification changes, and it can produce hex strings (of the 01-23-45 variety, but still one that people are used to).

Reply:
That format listed in the RFC doesn't have '-', following the same syntax from there.